### PR TITLE
refactor: move unwanted console output out of versiongetter API

### LINKED ISSUE #1853: Unwanted console output in the versiongetter API
API to get version information(ClusterVersion, KubeadmVersion and VersionFromCILabel) contain unwanted calls of fmt.*Print* functions:

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/versiongetter.go#L66
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/versiongetter.go#L78
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/versiongetter.go#L95

They should be moved outside of the API.  It should be done carefully in order to not degrade user experience, i.e. kubeadm console output should stay the same.

### PR DESCRIPTION
**Pull request by jfbai**
**What type of PR is this?**

/kind cleanup

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubeadm/issues/1853

**Special notes for your reviewer**:

None

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
None
```

/sig cluster-lifecycle
/assign @bart0sh
/assign @neolit123

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
@jfbai thanks.

i think the PR is doing more than needed, instead of changing function names, arguments and signatures it should simply move the fmt.Fprintf calls that @bart0sh listed here:
https://github.com/kubernetes/kubeadm/issues/1853#issue-511313003

on the side of all callers that need something to be printed to the user.

/hold

**Comment by jfbai** [comment]
> @jfbai thanks.
> 
> i think the PR is doing more than needed, instead of changing function names, arguments and signatures it should simply move the fmt.Fprintf calls that @bart0sh listed here:
> [kubernetes/kubeadm#1853 (comment)](https://github.com/kubernetes/kubeadm/issues/1853#issue-511313003)
> 
> on the side of all callers that need something to be printed to the user.
> 
> /hold

@neolit123 Thanks a lot. :)

Since there are `if` blocks before `fmt.Fprintf`, it would make the logic complicated on the caller side if we make sure that kubeadm console output should stay the same.

and there is a brief discussion about this solution. https://github.com/kubernetes/kubeadm/issues/1853#issuecomment-548427565

**Comment by neolit123** [comment]
> Since there are if blocks before fmt.Fprintf, it would make the logic complicated on the caller side if we make sure that kubeadm console output should stay the same.

if i recall correctly, currently the version strings are printed in places that we don't need them.

> and there is a brief discussion about this solution. kubernetes/kubeadm#1853 (comment)

i don't think we should split the API to suite the print calls.
printing on the caller side only in key places, makes more sense to me.

**Comment by jfbai** [comment]
@neolit123 Hi, I updated this PR based on your comment, PTAL, thanks a lot. :)

**Comment by neolit123** [comment]
@jfbai 
on a quick look this is much cleaner! i'm going to get back to this PR after code-freeze is lifted (22nd Nov):
https://github.com/kubernetes/sig-release/tree/master/releases/release-1.17#timeline

thank you!

**Comment by neolit123** [comment]
i'm ok with the current state of the change.
/approve

/assign @rosti 
PTAL for potential LGTM.

File: cmd/kubeadm/app/phases/upgrade/compute.go [hunk 1]
```diff
@@ -82,18 +82,25 @@ func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesA
 
 	// Get the cluster version
 	clusterVersionStr, clusterVersion, err := versionGetterImpl.ClusterVersion()
+	if len(clusterVersionStr) > 0 {
```
Comment (rosti) [inline cmd/kubeadm/app/phases/upgrade/compute.go]: In general, if `err != nil` we should not look at any other return values of funcs. Also, if there are no errors, these strings should be non-empty and printable (hence no checks are required).
Can you do the prints as else clauses to the error checks?
The same applies practically everywhere throughout this patch.

**Review by rosti — COMMENTED** [review]
Thanks @jfbai !

File: cmd/kubeadm/app/phases/upgrade/compute.go (hunk 1 above)
Comment (jfbai) [reply cmd/kubeadm/app/phases/upgrade/compute.go]: Fixed.

**Comment by neolit123** [comment]
@rosti PTAL another look for a LGTM.
/hold cancel

**Review by rosti — APPROVED** [review]
Thanks @jfbai !
Just a tiny nit more to cleanup. Can you remove the `io.Writer` from `KubeVersionGetter` and `NewKubeVersionGetter`?

**Comment by jfbai** [comment]
> Thanks @jfbai !
> Just a tiny nit more to cleanup. Can you remove the `io.Writer` from `KubeVersionGetter` and `NewKubeVersionGetter`?

@rosti Thanks, fixed.

**Review by rosti — APPROVED** [review]
Thanks @jfbai !
/lgtm

**Pull request merged**